File: problems/004/004.go
package main

import "fmt"
import "os"
import "strconv"
import "time"

func isPalindrome(n uint64) bool {
    reversed, tmp := uint64(0), n

    for tmp != 0 {
        reversed = 10 * reversed + (tmp % 10)
        tmp /= 10
    }

    return reversed == n
}

func f() uint64 {
    max := uint64(0)

    for i := uint64(0); i < 1000; i++ {
        for j := uint64(0); j < i; j++ {
            p := i * j

            if p > max && isPalindrome(p) {
                max = p
            }
        }
    }

    return max
}

func main() {
    if len(os.Args) == 1 {
        fmt.Println(f())
    } else if len(os.Args) == 2 {
        iters, _ := strconv.ParseInt(os.Args[1], 10, 64)

        start := time.Now()
        for i := int64(0); i < iters; i++ {
            f()
        }
        end := time.Now()

        fmt.Println(end.Sub(start).Nanoseconds())
    }
}

File: problems/004/004.rs
extern crate test;
extern crate time;

use std::os::args;
use test::black_box;
use time::precise_time_ns;

fn is_palindrome(n: u64) -> bool {
    let (mut reversed, mut tmp) = (0, n);

    while tmp != 0 {
        reversed = 10 * reversed + tmp % 10;
        tmp /= 10;
    }

    reversed == n
}

#[inline]
fn f() -> u64 {
    let mut max = 0;

    for a in range(100u64, 1000) {
        for b in range(100, a) {
            let p = a * b;

            if p > max && is_palindrome(p) {
                max = p;
            }
        }
    }

    max
}

fn main() {
    match args().as_slice() {
        [_] => {
            println!("{}", f());
        },
        [_, ref iters] => {
            let iters: u64 = from_str(iters.as_slice()).unwrap();

            let start = precise_time_ns();
            for _ in range(0, iters) {
                black_box(f());
            }
            let end = precise_time_ns();

            println!("{}", end - start);
        },
        _ => unreachable!(),
    }
}
